Habitica/src/main/java/com/habitrpg/android/habitica/models/notifications/GlobalNotification.kt

```diff
@@ -0,0 +1,64 @@
+package com.habitrpg.android.habitica.models.notifications
+
+import io.realm.RealmObject
+import io.realm.annotations.PrimaryKey
+import java.lang.reflect.Type
+
+enum class NotificationType(val type: String) {
+    NEW_STUFF("NEW_STUFF"),
+    NEW_CHAT_MESSAGE("NEW_CHAT_MESSAGE"),
+    UNALLOCATED_STATS_POINTS("UNALLOCATED_STATS_POINTS");
+
+    companion object {
+        fun contains(type: String?): Boolean {
+            return NotificationType.values().map { it.type }.contains(type)
+        }
+    }
+}
+
+/**
+ * Represents Habitica "Global notifications", i.e. the notifications about chat messages
+ * (in Guilds and Party), Party & Quest invitations, unallocated stat points etc.
+ *
+ * These are different from other kind of notifications, such as Push notifications and
+ * Popup notifications.
+ */
+open class GlobalNotification : RealmObject() {
+
+    @PrimaryKey
+    var id: String = ""
+
+    var type: String? = null
+    var seen: Boolean? = null
+
+    var newStuffData: NewStuffData? = null
+    var newChatMessageData: NewChatMessageData? = null
+    var unallocatedPointsData: UnallocatedPointsData? = null
+
+    // Workaround for Realms lack of polymorphism
+    fun getData(): GlobalNotificationData? {
+        return when (type) {
+            NotificationType.NEW_STUFF.type -> newStuffData
+            NotificationType.NEW_CHAT_MESSAGE.type -> newChatMessageData
+            NotificationType.UNALLOCATED_STATS_POINTS.type -> unallocatedPointsData
+            else -> null
```

Comment:
Not sure if there is better way to handle the `data` property of the notifications since it's different type for each notification (as Realm doesn't support polymorphism AFAIK). Suggestions/feedback is welcome. If I'm doing something stupid here, feel free to tell me :smiley:

Reply:
I'd maybe change the access level of the actual fields to something more restricted (since you're probably not supposed to interact with them directly), but while it does look strange it's probably the best solution given realms restrictions here!

Reply:
Looking at this again, I noticed that none of the Data objects have a primary key with the ID of the notification. This will be needed (or at least very helpful) for deleting old data from the database.

Reply:
Yeah, I'm actually currently looking into avoiding saving these to Realm at all. I realized at some point that the already existing `Notification` model is actually kind of same thing, just that it models one specific type of notification instead of all the different ones, and the `apiCallTransformer` in `ApiClientImpl` is something that handles these from each request. So I'm looking at if I could model all the notifications using similar approach as I did with `GlobalNotification` and then might hook into the apiCallTransformer as well with this new stuff.